src/MaplibreLayerRenderer.ts

```diff
@@ -0,0 +1,170 @@
+import {FrameState} from 'ol/Map';
+import {toDegrees} from 'ol/math';
+import {toLonLat} from 'ol/proj';
+import LayerRenderer from 'ol/renderer/Layer';
+import GeoJSON from 'ol/format/GeoJSON';
+import {Coordinate} from 'ol/coordinate';
+import {FeatureCallback} from 'ol/renderer/vector';
+import {Feature} from 'ol';
+import {Geometry, SimpleGeometry} from 'ol/geom';
+import {Pixel} from 'ol/pixel';
+import MapLibreLayer from './ol-maplibre-layer';
+
+const VECTOR_TILE_FEATURE_PROPERTY = 'vectorTileFeature';
+
+/**
+ * @private
+ */
+const formats: {
+  [key: string]: GeoJSON;
+} = {
+  'EPSG:3857': new GeoJSON({
+    featureProjection: 'EPSG:3857',
+  }),
+};
+
+/**
+ * This class is a renderer for Maplibre Layer to be able to use the native ol
+ * functionnalities like map.getFeaturesAtPixel or map.hasFeatureAtPixel.
+ * @private
+ */
+export default class MaplibreLayerRenderer extends LayerRenderer<MapLibreLayer> {
+  getFeaturesAtCoordinate(
+    coordinate: Coordinate | undefined,
+    hitTolerance: number = 5,
+  ): Feature<Geometry>[] {
+    if (!coordinate) {
+      return [];
+    }
+
+    const layer = this.getLayer();
+    const map = layer.getMapInternal();
+    const {maplibreMap} = layer;
+
+    const projection =
+      map?.getView()?.getProjection()?.getCode() || 'EPSG:3857';
+    let features: Feature[] = [];
+
+    if (!formats[projection]) {
+      formats[projection] = new GeoJSON({
+        featureProjection: projection,
+      });
+    }
+
+    if (maplibreMap) {
+      //?.isStyleLoaded()) {
+      const pixel =
+        coordinate &&
+        maplibreMap.project(toLonLat(coordinate) as [number, number]);
+
+      if (pixel?.x && pixel?.y) {
+        let pixels: [[number, number], [number, number]] | [number, number] = [
+          pixel.x,
+          pixel.y,
+        ];
+
+        if (hitTolerance) {
+          const [x, y] = pixels as [number, number];
+          pixels = [
+            [x - hitTolerance, y - hitTolerance],
+            [x + hitTolerance, y + hitTolerance],
+          ];
+        }
+
+        const queryRenderedFeaturesOptions =
+          (layer.get(
+            'queryRenderedFeaturesOptions',
+          ) as maplibregl.QueryRenderedFeaturesOptions) || {};
+
+        // At this point we get GeoJSON Maplibre feature, we transform it to an OpenLayers
+        // feature to be consistent with other layers.
+        features = maplibreMap
+          .queryRenderedFeatures(pixels, queryRenderedFeaturesOptions)
+          .map((feature) => {
+            const olFeature = formats[projection].readFeature(
+              feature,
+            ) as Feature;
+            if (olFeature) {
+              // We save the original Maplibre feature to avoid losing informations
+              // potentially needed for other functionnality like highlighting
+              // (id, layer id, source, sourceLayer ...)
+              olFeature.set(VECTOR_TILE_FEATURE_PROPERTY, feature);
```

Comment:
Maybe using the "silent" option of set could be great here. This method will be called a lot of time, and set can be heavy if events are triggered.
```suggestion
              olFeature.set(VECTOR_TILE_FEATURE_PROPERTY, feature, true);
```
What do you think ?

Reply:
Good idea yes